[AWQ] Evaluate fused vs unfused GEMM on actual shape

mgehre-amd opened the pull request:
to get a speed up of 2.5x for non-batched decode (single requests).

Before this PR, the condition
```
FP16_MATMUL_HEURISTIC_CONDITION = input.shape[0] >= 256
if FP16_MATMUL_HEURISTIC_CONDITION:
    out = awq_dequantize(qweight, scales, qzeros, 0, 0, 0)
    return torch.matmul(input, out)
else:
    return awq_gemm(...)
```
was evaluated during `torch.compile` based on `max-num-batched-tokens`. By default, `max-num-batched-tokens` is over the threshold, which meant that `awq_gemm` was never taken, even when later doing a single request decode (`input.shape[0] == 1`).

To evaluate the condition for each specific shape during during CUDA graph capture, this PR move the decision into `awq_gemm` and wraps that into a torch custom op, which shields it from being traced through.

Test results on Strix Halo (gfx1151):
```
$ vllm bench throughput --model Qwen/Qwen3-4B-AWQ --input-len 1024 \
     --gpu-memory-utilization 0.4 --max-model-len 4096 --num-prompts 30

# on main
Throughput: 0.75 requests/s, 860.06 total tokens/s, 95.56 output tokens/s
Total num prompt tokens:  30720
Total num output tokens:  3840

# on this PR
Throughput: 0.85 requests/s, 978.43 total tokens/s, 108.71 output tokens/s
Total num prompt tokens:  30720
Total num output tokens:  3840
```
Test results with `max-num-seq=1` on Strix Halo (gfx1151):
```
$ vllm bench throughput --model Qwen/Qwen3-4B-AWQ --input-len 1024 \
   --gpu-memory-utilization 0.4 --max-model-len 4096 --num-prompts 30 \
   --max-num-seqs=1

# on main
Throughput: 0.08 requests/s, 96.89 total tokens/s, 10.77 output tokens/s
Total num prompt tokens:  30720
Total num output tokens:  3840

# on this PR
Throughput: 0.20 requests/s, 231.17 total tokens/s, 25.69 output tokens/s
Total num prompt tokens:  30720
Total num output tokens:  3840
```

mgehre-amd committed:
AWQ: Evaluate fused vs unfused GEMM on actual shape

Before this PR, the condition
```
FP16_MATMUL_HEURISTIC_CONDITION = input.shape[0] >= 256
if FP16_MATMUL_HEURISTIC_CONDITION:
    out = awq_dequantize(qweight, scales, qzeros, 0, 0, 0)
    return torch.matmul(input, out)
else:
    return awq_gemm(...)
```
was evaluated during `torch.compile` based on `max-num-batched-tokens`.
By default, `max-num-batched-tokens` is over the threshold,
which meant that `awq_gemm` was never taken, even when doing a single
request decode.

To evaluate the condition for each specific shape during during
cudaGraph capture, this PR wraps `awq_gemm` into a torch custom op,
which shields it from being traced through.

Signed-off-by: Matthias Gehre <matthias.gehre@amd.com>

yuttian1 commented:
Hi, thanks for the PR. I might be missing some design context here, but I’m seeing the following error:
assert zeros.shape[0] == K // group_size and zeros.shape[1] == M // 8

mgehre-amd commented:
> assert zeros.shape[0] == K // group_size and zeros.shape[1] == M // 8

@yuttian1, thanks for your feedback.
I'm not sure how this could be caused by this PR. Can you please share 
1) the full commands to reproduce
2) the full output/traceback
3) and whether the same things works on vLLMs main branch?
Thanks!

yuttian1 commented:
> > assert zeros.shape[0] == K // group_size and zeros.shape[1] == M // 8
> 
> @yuttian1, thanks for your feedback. I'm not sure how this could be caused by this PR. Can you please share
> 
> 1. the full commands to reproduce
> 2. the full output/traceback
> 3. and whether the same things works on vLLMs main branch?
>    Thanks!
Sorry for the late reply！On my side, I’m running vLLM under ROCm.
Both awq_dequant + gemm and awq_gemm work fine when run independently, but this error appears after applying your changes.
the full output is
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "<eval_with_key>.190", line 1046, in forward
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     submod_0 = self.submod_0(l_self_modules_layers_modules_0_modules_input_layernorm_parameters_weight_, l_inputs_embeds_, s59, l_self_modules_layers_modules_0_modules_self_attn_modules_qkv_proj_parameters_qweight_, l_self_modules_layers_modules_0_modules_self_attn_modules_qkv_proj_parameters_scales_, l_self_modules_layers_modules_0_modules_self_attn_modules_qkv_proj_parameters_qzeros_, l_self_modules_layers_modules_0_modules_self_attn_modules_q_norm_parameters_weight_, l_self_modules_layers_modules_0_modules_self_attn_modules_k_norm_parameters_weight_, l_self_modules_layers_modules_0_modules_self_attn_modules_rotary_emb_buffers_cos_sin_cache_, l_positions_, s7);  l_self_modules_layers_modules_0_modules_input_layernorm_parameters_weight_ = l_self_modules_layers_modules_0_modules_self_attn_modules_qkv_proj_parameters_qweight_ = l_self_modules_layers_modules_0_modules_self_attn_modules_qkv_proj_parameters_scales_ = l_self_modules_layers_modules_0_modules_self_attn_modules_qkv_proj_parameters_qzeros_ = l_self_modules_layers_modules_0_modules_self_attn_modules_q_norm_parameters_weight_ = l_self_modules_layers_modules_0_modules_self_attn_modules_k_norm_parameters_weight_ = None
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/opt/vllm/vllm/compilation/cuda_graph.py", line 125, in __call__
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     return self.runnable(*args, **kwargs)
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/opt/vllm/vllm/compilation/piecewise_backend.py", line 92, in __call__
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     return self.compiled_graph_for_general_shape(*args)
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/usr/local/lib/python3.12/dist-packages/torch/_dynamo/eval_frame.py", line 850, in _fn
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     return fn(*args, **kwargs)
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/usr/local/lib/python3.12/dist-packages/torch/_functorch/aot_autograd.py", line 1207, in forward
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     return compiled_fn(full_args)
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/usr/local/lib/python3.12/dist-packages/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 331, in runtime_wrapper
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     all_outs = call_func_at_runtime_with_args(
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/usr/local/lib/python3.12/dist-packages/torch/_functorch/_aot_autograd/utils.py", line 126, in call_func_at_runtime_with_args
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     out = normalize_as_list(f(args))
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]                             ^^^^^^^
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/usr/local/lib/python3.12/dist-packages/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 692, in inner_fn
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     outs = compiled_fn(args)
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/usr/local/lib/python3.12/dist-packages/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 498, in wrapper
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     return compiled_fn(runtime_args)
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/usr/local/lib/python3.12/dist-packages/torch/_inductor/output_code.py", line 561, in __call__
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     return self.current_callable(inputs)
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/usr/local/lib/python3.12/dist-packages/torch/_inductor/utils.py", line 2444, in run
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     return model(new_inputs)
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/root/.cache/vllm/torch_compile_cache/ccf95ef384/rank_0_0/inductor_cache/2i/c2iqegvoziigca2hnpiv4jhaa4ywkmdsbcdatjahluc3c5gkscpz.py", line 321, in call
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     buf2 = torch.ops.vllm.awq_gemm.default(buf1, arg3_1, arg4_1, arg5_1, 8)
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/usr/local/lib/python3.12/dist-packages/torch/_ops.py", line 776, in __call__
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     return self._op(*args, **kwargs)
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/opt/vllm/vllm/_custom_ops.py", line 461, in _awq_gemm
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     out = awq_dequantize(qweight, scales, qzeros, 0, 0, 0)
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/opt/vllm/vllm/_custom_ops.py", line 442, in awq_dequantize
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     return awq_dequantize_triton(qweight, scales, zeros)
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/opt/vllm/vllm/model_executor/layers/quantization/awq_triton.py", line 246, in awq_dequantize_triton
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     assert zeros.shape[0] == K // group_size and zeros.shape[1] == M // 8
(EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

yuttian1 commented:
> > > assert zeros.shape[0] == K // group_size and zeros.shape[1] == M // 8
> > 
> > 
> > @yuttian1, thanks for your feedback. I'm not sure how this could be caused by this PR. Can you please share
> > 
> > 1. the full commands to reproduce
> > 2. the full output/traceback
> > 3. and whether the same things works on vLLMs main branch?
> >    Thanks!
> >    Sorry for the late reply！On my side, I’m running vLLM under ROCm.
> >    Both awq_dequant + gemm and awq_gemm work fine when run independently, but this error appears after applying your changes.
> >    the full output is
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "<eval_with_key>.190", line 1046, in forward
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     submod_0 = self.submod_0(l_self_modules_layers_modules_0_modules_input_layernorm_parameters_weight_, l_inputs_embeds_, s59, l_self_modules_layers_modules_0_modules_self_attn_modules_qkv_proj_parameters_qweight_, l_self_modules_layers_modules_0_modules_self_attn_modules_qkv_proj_parameters_scales_, l_self_modules_layers_modules_0_modules_self_attn_modules_qkv_proj_parameters_qzeros_, l_self_modules_layers_modules_0_modules_self_attn_modules_q_norm_parameters_weight_, l_self_modules_layers_modules_0_modules_self_attn_modules_k_norm_parameters_weight_, l_self_modules_layers_modules_0_modules_self_attn_modules_rotary_emb_buffers_cos_sin_cache_, l_positions_, s7);  l_self_modules_layers_modules_0_modules_input_layernorm_parameters_weight_ = l_self_modules_layers_modules_0_modules_self_attn_modules_qkv_proj_parameters_qweight_ = l_self_modules_layers_modules_0_modules_self_attn_modules_qkv_proj_parameters_scales_ = l_self_modules_layers_modules_0_modules_self_attn_modules_qkv_proj_parameters_qzeros_ = l_self_modules_layers_modules_0_modules_self_attn_modules_q_norm_parameters_weight_ = l_self_modules_layers_modules_0_modules_self_attn_modules_k_norm_parameters_weight_ = None
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/opt/vllm/vllm/compilation/cuda_graph.py", line 125, in **call**
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     return self.runnable(*args, **kwargs)
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/opt/vllm/vllm/compilation/piecewise_backend.py", line 92, in **call**
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     return self.compiled_graph_for_general_shape(*args)
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/usr/local/lib/python3.12/dist-packages/torch/_dynamo/eval_frame.py", line 850, in _fn
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     return fn(*args, **kwargs)
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^^^
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/usr/local/lib/python3.12/dist-packages/torch/_functorch/aot_autograd.py", line 1207, in forward
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     return compiled_fn(full_args)
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^^^^^^
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/usr/local/lib/python3.12/dist-packages/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 331, in runtime_wrapper
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     all_outs = call_func_at_runtime_with_args(
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/usr/local/lib/python3.12/dist-packages/torch/_functorch/_aot_autograd/utils.py", line 126, in call_func_at_runtime_with_args
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     out = normalize_as_list(f(args))
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]                             ^^^^^^^
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/usr/local/lib/python3.12/dist-packages/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 692, in inner_fn
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     outs = compiled_fn(args)
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/usr/local/lib/python3.12/dist-packages/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 498, in wrapper
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     return compiled_fn(runtime_args)
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^^^^^^^^^
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/usr/local/lib/python3.12/dist-packages/torch/_inductor/output_code.py", line 561, in **call**
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     return self.current_callable(inputs)
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/usr/local/lib/python3.12/dist-packages/torch/_inductor/utils.py", line 2444, in run
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     return model(new_inputs)
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/root/.cache/vllm/torch_compile_cache/ccf95ef384/rank_0_0/inductor_cache/2i/c2iqegvoziigca2hnpiv4jhaa4ywkmdsbcdatjahluc3c5gkscpz.py", line 321, in call
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     buf2 = torch.ops.vllm.awq_gemm.default(buf1, arg3_1, arg4_1, arg5_1, 8)
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/usr/local/lib/python3.12/dist-packages/torch/_ops.py", line 776, in **call**
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     return self._op(*args, **kwargs)
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^^^^^^^^^
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/opt/vllm/vllm/_custom_ops.py", line 461, in _awq_gemm
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     out = awq_dequantize(qweight, scales, qzeros, 0, 0, 0)
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/opt/vllm/vllm/_custom_ops.py", line 442, in awq_dequantize
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     return awq_dequantize_triton(qweight, scales, zeros)
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]   File "/opt/vllm/vllm/model_executor/layers/quantization/awq_triton.py", line 246, in awq_dequantize_triton
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]     assert zeros.shape[0] == K // group_size and zeros.shape[1] == M // 8
> >    (EngineCore_DP0 pid=12538) ERROR 12-19 06:40:14 [core.py:790]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

I’ve looked into this further and found the issue was on my side. My local vLLM version was outdated, and the argument order of scales and qzeros didn’t match the current implementation. After fixing that, the issue no longer reproduces.

With this PR applied, I do see a performance improvement. Under my test setup (concurrency = 1, ~88k input tokens), I observe performance improvement!

mgehre-amd committed:
Fix _awq_gemm_fake_impl dtype to be consistent with CUDA kernel

Use input.dtype instead of scales.dtype to match the underlying
_C::awq_gemm fake implementation for torch.compile type inference.

Signed-off-by: Matthias Gehre <matthias.gehre@amd.com>

mgehre-amd committed:
Merge remote-tracking branch 'upstream/main' into matthias.awq_specialize

Signed-off-by: Matthias Gehre <matthias.gehre@amd.com>
